### PR TITLE
Create interactor to generate syncable data dump

### PR DESCRIPTION
**Pull request by daividssilverio**
## What's this?
This is one of the interactors that will build the data dump of unsynced entities that will end up in a file to be consumed in the new apps.

More details and how everything else can/will be done can be found at #7104.

### Relationships
Part of #7104

## Why do we want this?
With this interactor, we have the first part needed to create the dump files: The data we'll need.
The `SyncableDataDump` can then be serialized into json and an actual file can be created in both platforms using whatever is decided at #7104.

## How is it done?
The `CreateUnsyncedDataDumpInteractor` takes in the toggl data sources and spits out a `SyncableDataDump`, which contains a list of `TimeEntries`, `Tags`, `Projects` and `Clients` which have their sync status in either `SyncNeeded`, `SyncFailed` or `RefetchingNeeded`.

### Why not another way?
By all means, suggest something different if think this won't work.

### Side effects
None yet, we are not running this yet.

## Review hints
Check the unit tests. 
You setup some offline testing and run this interactor artificially in some view model just to see it working as well.

## :squid: Permissions
🦑 it

### CODE REVIEW

File: Toggl.Core/Models/SyncableDataDump.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+using Toggl.Core.Models.Interfaces;
+
+namespace Toggl.Core.Models
+{
+    public class SyncableDataDump
+    {
+        public IEnumerable<IThreadSafeTimeEntry> TimeEntries { get; }
+        public IEnumerable<IThreadSafeTag> Tags { get; }
+        public IEnumerable<IThreadSafeProject> Projects { get; }
+        public IEnumerable<IThreadSafeClient> Clients { get; }
```
Comment (heytherewill) [inline Toggl.Core/Models/SyncableDataDump.cs]: Make these `IImmutableList<IThreadSafeTimeEntry>` to ensure that the lists are enumerated and can't be further changed

File: Toggl.Core/Models/SyncableDataDump.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+        public SyncableDataDump(
+            IEnumerable<IThreadSafeTimeEntry> timeEntries,
+            IEnumerable<IThreadSafeTag> tags,
+            IEnumerable<IThreadSafeProject> projects,
+            IEnumerable<IThreadSafeClient> clients)
```
Comment (heytherewill) [inline Toggl.Core/Models/SyncableDataDump.cs]: You can still receive `IEnumerable<T>` here, just make sure to force enumeration by calling `ToImmutableList`

File: Toggl.Core/Interactors/Sync/CreateUnsyncedDataDumpInteractor.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            return new SyncableDataDump(timeEntries, tags, projects, clients);
+        }
+
+        private IObservable<IEnumerable<TThreadsafe>> getUnsyncedItems<TThreadsafe, TDatabase>(
+            IDataSource<TThreadsafe, TDatabase> source)
+            where TDatabase : IDatabaseSyncable
+            where TThreadsafe : TDatabase, IThreadSafeModel
+        {
+            return source.GetAll(p => p.SyncStatus != SyncStatus.InSync);
+        }
```
Comment (heytherewill) [inline Toggl.Core/Interactors/Sync/CreateUnsyncedDataDumpInteractor.cs]: EBM wouldn't hurt here:
```suggestion
        => source.GetAll(p => p.SyncStatus != SyncStatus.InSync);
```

File: Toggl.Core/Interactors/Sync/CreateUnsyncedDataDumpInteractor.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            Ensure.Argument.IsNotNull(dataSource, nameof(dataSource));
+            this.dataSource = dataSource;
+        }
+
+        public async Task<SyncableDataDump> Execute()
+        {
+            var clients = await getUnsyncedItems(dataSource.Clients);
+            var projects = await getUnsyncedItems(dataSource.Projects);
+            var tags = await getUnsyncedItems(dataSource.Tags);
+            var timeEntries = await getUnsyncedItems(dataSource.TimeEntries);
```
Comment (heytherewill) [inline Toggl.Core/Interactors/Sync/CreateUnsyncedDataDumpInteractor.cs]:
```suggestion
            var clients = await dataSource.Clients.GetAll(notSynced);
            var projects = await dataSource.Projects.GetAll(notSynced);
            var tags = await dataSource.Tags.GetAll(notSynced);
            var timeEntries = await dataSource.TimeEntries.GetAll(notSynced);
```

Comment (heytherewill) [reply Toggl.Core/Interactors/Sync/CreateUnsyncedDataDumpInteractor.cs]: This is actually shorter 😂

File: Toggl.Core/Models/SyncableDataDump.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
+using System.Collections.Generic;
+using Toggl.Core.Models.Interfaces;
+
+namespace Toggl.Core.Models
+{
+    public class SyncableDataDump
```
Comment (juanlaube) [inline Toggl.Core/Models/SyncableDataDump.cs]: Shouldn't this be `UnsyncableDataDump`? 😄

Comment (juanlaube) [reply Toggl.Core/Models/SyncableDataDump.cs]: Wouldn't hurt to add a method in this class to serialize that dump into JSON so the next parts of this issue can do the writing only.

Comment (heytherewill) [reply Toggl.Core/Models/SyncableDataDump.cs]: Unsyncable is not accurate either; `UnsyncedDataDump` is more adequate

Comment (daividssilverio) [reply Toggl.Core/Models/SyncableDataDump.cs]: Not really, right? This is the the dump of the data that couldn't be synced, not the data that is unsyncable.

Maybe `UnsyncedDataDump` would make more sense?

Comment (juanlaube) [reply Toggl.Core/Models/SyncableDataDump.cs]: You are right, `UnsyncedDataDump` is 💯

Comment (daividssilverio) [reply Toggl.Core/Models/SyncableDataDump.cs]: @juanlaube I think the serialization can be done in another PR before or alongside the services for each platform.

**Pull request merged**